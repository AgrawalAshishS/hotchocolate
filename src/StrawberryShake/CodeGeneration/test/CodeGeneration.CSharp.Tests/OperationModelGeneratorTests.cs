using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Snapshooter.Xunit;
using Xunit;

namespace StrawberryShake.CodeGeneration.CSharp
{
    public class OperationModelGeneratorTests
    {
        [Fact]
        public async Task GenerateClient_With_Singel_Optional_Argument()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new OperationModelGenerator();

            var descriptor = new OperationModelDescriptor(
                "OnReviewOperation",
                "Demo",
                "onReview",
                "global::Demo.IOnReview",
                "Queries",
                "Subscription",
                new List<OperationArgumentDescriptor>
                {
                    new OperationArgumentDescriptor(
                        "Episode",
                        "episode",
                        "episode",
                        "global::StrawberryShake.Optional<global::Demo.Episode>",
                        "Episode",
                        true)
                });

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task GenerateClient_With_No_Argument()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new OperationModelGenerator();

            var descriptor = new OperationModelDescriptor(
                "OnReviewOperation",
                "Demo",
                "onReview",
                "global::Demo.IOnReview",
                "Queries",
                "Subscription",
                new List<OperationArgumentDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task GenerateClient_With_Multiple_Arguments()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new OperationModelGenerator();

            var descriptor = new OperationModelDescriptor(
                "OnReviewOperation",
                "Demo",
                "onReview",
                "global::Demo.IOnReview",
                "Queries",
                "Subscription",
                new List<OperationArgumentDescriptor>
                {
                    new OperationArgumentDescriptor(
                        "Episode",
                        "episode",
                        "episode",
                        "global::StrawberryShake.Optional<global::Demo.Episode>",
                        "Episode",
                        true),
                    new OperationArgumentDescriptor(
                        "Abc",
                        "abcd",
                        "abc",
                        "string",
                        "String",
                        false),
                    new OperationArgumentDescriptor(
                        "Def",
                        "defg",
                        "def",
                        "int",
                        "Int",
                        false)
                });

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public void CanHandle()
        {
            // arrange
            var generator = new OperationModelGenerator();

            var descriptor = new OperationModelDescriptor(
                "OnReviewOperation",
                "Demo",
                "onReview",
                "global::Demo.IOnReview",
                "Queries",
                "Subscription",
                new List<OperationArgumentDescriptor>());

            // act
            var canHandle = generator.CanHandle(descriptor);

            // assert
            Assert.True(canHandle);
        }
    }
}
