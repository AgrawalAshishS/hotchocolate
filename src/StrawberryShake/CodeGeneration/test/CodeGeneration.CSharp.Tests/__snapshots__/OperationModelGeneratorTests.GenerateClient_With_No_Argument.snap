namespace Demo
{
    [global::System.CodeDom.Compiler.GeneratedCode("StrawberryShake", "11.0.0")]
    public partial OnReviewOperation
        : global::StrawberryShake.IOperation<global::Demo.IOnReview>
    {
        public string Name
        {
            get
            {
                return "onReview";
            }
        }

        public global::StrawberryShake.IDocument Document
        {
            get
            {
                return Queries.Default;
            }
        }

        public global::StrawberryShake.OperationKind Kind
        {
            get
            {
                return OperationKind.Subscription;
            }
        }

        public global::System.Type ResultType
        {
            get
            {
                return typeof(global::Demo.IOnReview);
            }
        }

        public global::System.Collections.Generic.IReadOnlyList<global::StrawberryShake.VariableValue> GetVariableValues()
        {
            return global::System.Array.Empty<global::StrawberryShake.VariableValue>();
        }
    }
}
